### PR TITLE
Align pareto front graph between optuna and optuna-dashboard

### PR DESCRIPTION
**Pull request by keisuke-umezawa**
## Contributor License Agreement

This repository (``optuna-dashboard``) and [Goptuna](https://github.com/c-bata/goptuna) share common code.
This pull request may therefore be ported to Goptuna.
Make sure that you understand the consequences concerning licenses and check the box below if you accept the term before creating this pull request.

* [x] I agree this patch may be ported to [Goptuna](https://github.com/c-bata/goptuna) by other Goptuna contributors.

## Reference Issues/PRs

None

## What does this implement/fix? Explain your changes.

Currently, there are some differences between optuna and optuna-dashboard about plotting optimization history as follows:
* "hovertemplate":"%{text}<extra>Trial</extra>"
* Different color scale of trial and best trial
* "showlegend":false,
* text

| Target (optuna) | Master (optuna-dashboard) | PR (optuna-dashboard) |
|:---|:---:|---:|
|plot|newplot (2)|newplot (2)|

* Target (optuna)
```
[
   {
      "hovertemplate":"%{text}<extra>Trial</extra>",
      "marker":{
         "color":[
            1,
            3,
            7,
            8
         ],
         "colorbar":{
            "title":{
               "text":"Trial"
            }
         },
         "colorscale":"Blues",
         "line":{
            "color":"Grey",
            "width":0.5
         }
      },
      "mode":"markers",
      "showlegend":false,
      "text":[
         "{<br>  \"number\": 1,<br>  \"values\": [<br>    ""16.902275647775767,<br>    2.390944855859108<br>  ],<br>  ""\"params\": {<br>    \"x1\": 1.4675589081711304,<br>    \"x2\": ""0.923385947687978,<br>    \"x3\": \"bar\"<br>  }<br>}",
         "{<br>  \"number\": 3,<br>  \"values\": [<br>    ""14.299261823053241,<br>    10.825696861224628<br>  ],<br>  ""\"params\": {<br>    \"x1\": 2.0445224973151745,<br>    \"x2\": ""8.781174363909454,<br>    \"x3\": \"bar\"<br>  }<br>}",
         "{<br>  \"number\": 7,<br>  \"values\": [<br>    ""14.390195648351362,<br>    10.479729229939819<br>  ],<br>  ""\"params\": {<br>    \"x1\": 1.698304195645689,<br>    \"x2\": ""8.78142503429413,<br>    \"x3\": \"bar\"<br>  }<br>}",
         "{<br>  \"number\": 8,<br>  \"values\": [<br>    ""57.549647603909,<br>    14.91054815123519<br>  ],<br>  \"params\": ""{<br>    \"x1\": 9.57889530150502,<br>    \"x2\": ""5.331652849730171,<br>    \"x3\": \"foo\"<br>  }<br>}"
      ],
      "type":"scatter",
      "x":[...],
      "y":[...]
   },
   {
      "hovertemplate":"%{text}<extra>Best Trial</extra>",
      "marker":{
         "color":[
            0,
            2,
            4,
            5,
            6,
            9
         ],
         "colorbar":{
            "title":{
               "text":"Best Trial"
            },
            "x":1.1,
            "xpad":40
         },
         "colorscale": "Reds",
         "line":{
            "color":"Grey",
            "width":0.5
         }
      },
      "mode":"markers",
      "showlegend":false,
      "text":[
         "{<br>  \"number\": 0,<br>  \"values\": [<br>    ""9.564143293566762,<br>    11.37346498144732<br>  ],<br>  \"params\": ""{<br>    \"x1\": 4.17022004702574,<br>    \"x2\": ""7.203244934421581,<br>    \"x3\": \"bar\"<br>  }<br>}",
         "{<br>  \"number\": 2,<br>  \"values\": [<br>    ""4.022417775380473,<br>    9.355842082340269<br>  ],<br>  \"params\": ""{<br>    \"x1\": 3.9676747423066994,<br>    \"x2\": ""5.3881673400335695,<br>    \"x3\": \"bar\"<br>  }<br>}",
         "{<br>  \"number\": 4,<br>  \"values\": [<br>    ""5.066587309480789,<br>    9.759946308128786<br>  ],<br>  \"params\": ""{<br>    \"x1\": 4.17304802367127,<br>    \"x2\": ""5.586898284457517,<br>    \"x3\": \"bar\"<br>  }<br>}",
         "{<br>  \"number\": 5,<br>  \"values\": [<br>    ""58.01629400883696,<br>    17.69006144394934<br>  ],<br>  \"params\": ""{<br>    \"x1\": 8.007445686755366,<br>    \"x2\": ""9.682615757193975,<br>    \"x3\": \"bar\"<br>  }<br>}",
         "{<br>  \"number\": 6,<br>  \"values\": [<br>    ""61.3216704225392,<br>    17.709958157998855<br>  ],<br>  \"params\": ""{<br>    \"x1\": 8.763891522960384,<br>    \"x2\": ""8.946066635038473,<br>    \"x3\": \"foo\"<br>  }<br>}",
         "{<br>  \"number\": 9,<br>  \"values\": [<br>    ""34.865749292780976,<br>    15.211265995789566<br>  ],<br>  ""\"params\": {<br>    \"x1\": 6.865009276815837,<br>    \"x2\": ""8.346256718973729,<br>    \"x3\": \"bar\"<br>  }<br>}"
      ],
      "type":"scatter",
      "x":[...],
      "y":[...]
   }
]
```

* Master (optuna-dashboard)
```
[
   {
      "type":"scatter",
      "x":[
         9.564143293566762,
         16.902275647775767,
         4.022417775380473,
         14.299261823053241,
         5.066587309480789,
         58.01629400883696,
         61.3216704225392,
         14.390195648351362,
         57.549647603909,
         34.865749292780976
      ],
      "y":[
         11.37346498144732,
         2.390944855859108,
         9.355842082340269,
         10.825696861224628,
         9.759946308128786,
         17.69006144394934,
         17.709958157998855,
         10.479729229939819,
         14.91054815123519,
         15.211265995789566
      ],
      "mode":"markers",
      "marker":{
         "color":[
            "red",
            "blue",
            "red",
            "blue",
            "red",
            "red",
            "red",
            "blue",
            "blue",
            "red"
         ]
      },
      "text":[
         "Trial (number=0)",
         "Trial (number=1)",
         "Trial (number=2)",
         "Trial (number=3)",
         "Trial (number=4)",
         "Trial (number=5)",
         "Trial (number=6)",
         "Trial (number=7)",
         "Trial (number=8)",
         "Trial (number=9)"
      ],
      "hovertemplate":"%{text}<extra></extra>"
   }
]
```

* PR (optuna-dashboard)
```
[
   {
      "x":[...],
      "y":[...],
      "text":[
         "{<br>  \"number\": 1,<br>  \"values\": [<br>    16.902275647775767,<br>    2.390944855859108<br>  ],<br>  \"params\": {<br>    \"x1\": \"1.4675589081711304\",<br>    \"x2\": \"0.923385947687978\",<br>    \"x3\": \"bar\"<br>  }<br>}",
         "{<br>  \"number\": 3,<br>  \"values\": [<br>    14.299261823053241,<br>    10.825696861224628<br>  ],<br>  \"params\": {<br>    \"x1\": \"2.0445224973151745\",<br>    \"x2\": \"8.781174363909454\",<br>    \"x3\": \"bar\"<br>  }<br>}",
         "{<br>  \"number\": 7,<br>  \"values\": [<br>    14.390195648351362,<br>    10.479729229939819<br>  ],<br>  \"params\": {<br>    \"x1\": \"1.698304195645689\",<br>    \"x2\": \"8.78142503429413\",<br>    \"x3\": \"bar\"<br>  }<br>}",
         "{<br>  \"number\": 8,<br>  \"values\": [<br>    57.549647603909,<br>    14.91054815123519<br>  ],<br>  \"params\": {<br>    \"x1\": \"9.57889530150502\",<br>    \"x2\": \"5.331652849730171\",<br>    \"x3\": \"foo\"<br>  }<br>}"
      ],
      "mode":"markers",
      "hovertemplate":"%{text}<extra>Trial</extra>",
      "marker":{
         "line":{
            "width":0.5,
            "color":"Grey"
         },
         "color":[
            1,
            3,
            7,
            8
         ],
         "colorscale":"Blues",
         "reversescale":true,
         "colorbar":{
            "title":{
               "text":"Trial"
            }
         }
      },
      "showlegend":false
   },
   {
      "x":[...],
      "y":[...],
      "text":[
         "{<br>  \"number\": 0,<br>  \"values\": [<br>    9.564143293566762,<br>    11.37346498144732<br>  ],<br>  \"params\": {<br>    \"x1\": \"4.17022004702574\",<br>    \"x2\": \"7.203244934421581\",<br>    \"x3\": \"bar\"<br>  }<br>}",
         "{<br>  \"number\": 2,<br>  \"values\": [<br>    4.022417775380473,<br>    9.355842082340269<br>  ],<br>  \"params\": {<br>    \"x1\": \"3.9676747423066994\",<br>    \"x2\": \"5.3881673400335695\",<br>    \"x3\": \"bar\"<br>  }<br>}",
         "{<br>  \"number\": 4,<br>  \"values\": [<br>    5.066587309480789,<br>    9.759946308128786<br>  ],<br>  \"params\": {<br>    \"x1\": \"4.17304802367127\",<br>    \"x2\": \"5.586898284457517\",<br>    \"x3\": \"bar\"<br>  }<br>}",
         "{<br>  \"number\": 5,<br>  \"values\": [<br>    58.01629400883696,<br>    17.69006144394934<br>  ],<br>  \"params\": {<br>    \"x1\": \"8.007445686755366\",<br>    \"x2\": \"9.682615757193975\",<br>    \"x3\": \"bar\"<br>  }<br>}",
         "{<br>  \"number\": 6,<br>  \"values\": [<br>    61.3216704225392,<br>    17.709958157998855<br>  ],<br>  \"params\": {<br>    \"x1\": \"8.763891522960384\",<br>    \"x2\": \"8.946066635038473\",<br>    \"x3\": \"foo\"<br>  }<br>}",
         "{<br>  \"number\": 9,<br>  \"values\": [<br>    34.865749292780976,<br>    15.211265995789566<br>  ],<br>  \"params\": {<br>    \"x1\": \"6.865009276815837\",<br>    \"x2\": \"8.346256718973729\",<br>    \"x3\": \"bar\"<br>  }<br>}"
      ],
      "mode":"markers",
      "hovertemplate":"%{text}<extra>Best Trial</extra>",
      "marker":{
         "line":{
            "width":0.5,
            "color":"Grey"
         },
         "color":[
            0,
            2,
            4,
            5,
            6,
            9
         ],
         "colorscale":"Reds",
         "colorbar":{
            "title":{
               "text":"Best Trial"
            },
            "x":1.1,
            "xpad":80,
            "y":0.5
         }
      },
      "showlegend":false
   }
]
```
By this PR, I removed the differences.

## Script
```
import pprint

import optuna
from optuna_dashboard import wsgi


def main():
    storage = optuna.storages.InMemoryStorage()
    sampler = optuna.samplers.RandomSampler(seed=1)
    study = optuna.create_study(
        study_name="single-objective",
        directions=["minimize", "maximize"],
        storage=storage,
        sampler=sampler,
    )

    def objective_single(trial: optuna.Trial) -> float:
        x1 = trial.suggest_float("x1", 0, 10)
        x2 = trial.suggest_float("x2", 0, 10)
        x3 = trial.suggest_categorical("x3", ["foo", "bar"])
        return (x1 - 2) ** 2 + (x2 - 5) ** 2, x1 + x2

    study.optimize(objective_single, n_trials=10)

    fig = optuna.visualization.plot_pareto_front(
        study,
    )
    fig.update_layout(
        width=800,
        height=600,
    )
    print("")
    print("Data")
    pprint.pprint(fig._data)
    print("")
    print("Layout")
    pprint.pprint(fig._layout)
    fig.write_image("plot.png")

    app = wsgi(storage)
    app.run(port=8080)


if __name__ == '__main__':
    main()
```

### CODE REVIEW

**Comment by keisuke-umezawa** [comment]
@c-bata Could you review it when you have time?

**Comment by c-bata** [comment]
@keisuke-umezawa Thanks for the update! Could you rebase on main branch?

**Comment by keisuke-umezawa** [comment]
@c-bata Rebased against main.

File: optuna_dashboard/ts/components/GraphParetoFront.tsx [hunk 1]
```diff
@@ -94,6 +94,69 @@ const filterFunc = (trial: Trial, directions: StudyDirection[]): boolean => {
...
+  } else {
+    return {
+      line: { width: 0.5, color: "Grey" },
+      // @ts-ignore
+      color: trials.map((t) => t.number),
+      colorscale: "Reds",
+      colorbar: {
+        title: "Best Trial",
+        x: 1.1,
+        xpad: 80,
```
Comment (c-bata) [inline optuna_dashboard/ts/components/GraphParetoFront.tsx:154]: [nit] It looks that Optuna uses `xpad: 40`. Do you think we should use 80 here?

**Review by c-bata — COMMENTED** [review]
Looks almost good to me. I left one question though.

File: optuna_dashboard/ts/components/GraphParetoFront.tsx (hunk 1 above)
Comment (keisuke-umezawa) [reply optuna_dashboard/ts/components/GraphParetoFront.tsx:154]: @c-bata 
Yes. Otherwise, the red and blue bars will be overlapped.

Comment (c-bata) [reply optuna_dashboard/ts/components/GraphParetoFront.tsx:154]: I see. Thanks. LGTM!

**Review by c-bata — APPROVED** [review]

**Pull request merged**